src/jshint.js

```diff
@@ -5032,115 +4831,625 @@ var JSHINT = (function() {
+         * see current.add for block scoped
+         */
+        add: function(labelName, type, tok, unused) {
+          _current["(labels)"][labelName] = {
+            "(type)" : type,
+            "(token)": tok,
+            "(blockscoped)": false,
+            "(unused)": unused };
         }
       },
 
       current: {
-        labeltype: function(t) {
-          if (_current[t]) {
-            return _current[t]["(type)"];
+        use: function(labelName, token) {
+
+          // if resolves to current function params, then do not store usage just resolve
+          // this is because function(a) { var a = a; } will resolve to the param, not
+          // to the unset var
+          // first check the param is used
+          var paramScope = _currentFunct["(parent)"];
+          if (paramScope && _.has(paramScope["(labels)"], labelName) &&
+            paramScope["(labels)"][labelName]["(type)"] === "param") {
+
+            // then check its not used
+            if (!funct["(scope)"].functionScope.has(labelName, { excludeParams: true })) {
+              paramScope["(labels)"][labelName]["(unused)"] = false;
+            }
           }
-          return null;
+
+          if (token && (state.ignored.W117 || state.option.undef === false)) {
+            token.ignoreUndef = true;
+          }
+
+          _addUsage(labelName, token);
+        },
+
+        reassigned: function(labelName, token) {
+
+          this.modify(labelName, token);
+
+          _current["(usages)"][labelName]["(reassigned)"].push(token);
         },
 
-        has: function(t) {
-          return _.has(_current, t);
+        modify: function(labelName, token) {
+
+          _current["(usages)"][labelName]["(modified)"].push(token);
         },
 
-        add: function(t, type, tok) {
-          _current[t] = { "(type)" : type, "(token)": tok, "(shadowed)": false, "(unused)": true };
+        /**
+         * Adds a new variable
+         */
+        add: function(labelName, type, tok, unused) {
+          _current["(labels)"][labelName] = {
+            "(type)" : type,
+            "(token)": tok,
+            "(blockscoped)": true,
+            "(unused)": unused };
+        },
+
+        addBreakLabel: function(labelName, opts) {
+          var token = opts.token;
+          if (state.option.shadow === "outer") {
+            if (funct["(scope)"].functionScope.has(labelName)) {
+              warning("W004", token, labelName);
+            } else {
+              _checkOuterShadow(labelName, token);
+            }
+          }
+          _current["(breakLabels)"][labelName] = token;
+        },
+
+        // adds an indentifier to the relevant current scope and creates warnings/errors as necessary
+        // name: string
+        // opts: { type: string, token: token, isblockscoped: bool }
+        addlabel: function(labelName, opts) {
+
+          var type  = opts.type;
+          var token = opts.token;
+          var isblockscoped = opts.isblockscoped;
```

Comment:
yes I can now. when we were adding break labels to the same object, they got true too. I think the only other one is the catch param, but that gets added through stackParams, so will do this.